src/passes/implicitConversionToExplicit.ts

```diff
@@ -257,34 +256,42 @@ export class ImplicitConversionToExplicit extends ASTMapper {
 }
 
 function insertConversionIfNecessary(expression: Expression, targetType: TypeNode, ast: AST): void {
-  const currentType = generalizeType(getNodeType(expression, ast.compilerVersion))[0];
-  targetType = generalizeType(targetType)[0];
+  const [currentType, currentLoc] = generalizeType(getNodeType(expression, ast.compilerVersion));
+  const generalisedTargetType = generalizeType(targetType)[0];
 
   if (currentType instanceof AddressType) {
-    if (!(targetType instanceof AddressType)) {
-      insertConversion(expression, targetType, ast);
+    if (!(generalisedTargetType instanceof AddressType)) {
+      insertConversion(expression, generalisedTargetType, ast);
     }
   } else if (currentType instanceof ArrayType) {
     assert(
-      targetType instanceof ArrayType,
+      generalisedTargetType instanceof ArrayType,
       `Unable to convert array ${printNode(expression)} to non-array type ${printTypeNode(
-        targetType,
+        generalisedTargetType,
       )}`,
     );
-    const elementT = targetType.elementT;
-    if (expression instanceof TupleExpression && expression.isInlineArray) {
-      expression.vOriginalComponents.forEach((element) => {
-        assert(element !== null, `Unexpected empty slot in inline array ${printNode(expression)}`);
-        insertConversionIfNecessary(element, elementT, ast);
-      });
-      expression.typeString = generateExpressionTypeString(
-        specializeType(targetType, DataLocation.Memory),
-      );
+    if (currentLoc === DataLocation.Memory) {
+      const parent = expression.parent;
+      const [replacement, shouldReplace] = ast
+        .getUtilFuncGen(expression)
+        .memory.convert.genIfNecesary(expression, targetType);
+      if (shouldReplace) {
+        ast.replaceNode(expression, replacement, parent);
+      }
     }
+    // if (expression instanceof TupleExpression && expression.isInlineArray) {
```

Comment:
Commented code

Reply:
This aws actually on purpose this time because this is a potential optimisation, I can remove it if you'd like though

Reply:
Nope no worries.